pkg/placement/ha_test.go

```diff
@@ -0,0 +1,193 @@
+package placement
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/dapr/dapr/pkg/placement/raft"
+)
+
+func TestPlacementHA(t *testing.T) {
+	testMembers := []*raft.DaprHostMember{
+		{
+			Name:     "127.0.0.1:3031",
+			AppID:    "testmember1",
+			Entities: []string{"red"},
+		},
+		{
+			Name:     "127.0.0.1:3032",
+			AppID:    "testmember2",
+			Entities: []string{"blue"},
+		},
+		{
+			Name:     "127.0.0.1:3033",
+			AppID:    "testmember3",
+			Entities: []string{"red", "blue"},
+		},
+	}
+
+	// Create 3 raft servers
+	raftServers := make([]*raft.Server, 3)
+	peers := make([]raft.PeerInfo, 3)
+	for i := 0; i < 3; i++ {
+		peers[i] = raft.PeerInfo{
+			ID:      fmt.Sprintf("mynode-%d", i),
+			Address: fmt.Sprintf("127.0.0.1:%d", 6100+i),
+		}
+	}
+	for i := 0; i < 3; i++ {
+		raftServers[i] = createRaftServer(t, i, peers)
+	}
+
+	// Run tests
+	leader := -1
+	t.Run("elects leader with 3 nodes", func(t *testing.T) {
+		leader = findLeader(t, raftServers)
+	})
+	if leader == -1 {
+		// If there's no leader, no point in running the other tests
+		t.FailNow()
+	}
+	t.Run("set and retrieve state in leader", func(t *testing.T) {
+		_, err := raftServers[leader].ApplyCommand(raft.MemberUpsert, *testMembers[0])
+		assert.NoError(t, err)
+
+		retrieveValidState(t, raftServers[leader], testMembers[0])
+	})
+	t.Run("retrieve state in follower", func(t *testing.T) {
+		time.Sleep(2500 * time.Millisecond)
+		follower := (leader + 1) % 3
+		retrieveValidState(t, raftServers[follower], testMembers[0])
+	})
+	t.Run("new leader after leader fails", func(t *testing.T) {
+		raftServers[leader].Shutdown()
+		raftServers[leader] = nil
+		time.Sleep(2500 * time.Millisecond)
+		leader = findLeader(t, raftServers)
+		raftServers[leader].FSM().State()
+	})
+	t.Run("set and retrieve state in leader", func(t *testing.T) {
```

Comment:
Can this have a unique name to distinguish it from the previous test case? "set and retrieve state in leader after reelection"?

Reply:
Fixed